Centralize opcode enum

oneturkmen opened the pull request:
Issue: #4519 

Refactoring one header file at a time.

**Optimizer**

- [x] compiler/optimizer/SimplifierTable.hpp
- [x] compiler/optimizer/ValuePropagationTable.hpp
- [x] compiler/optimizer/OMRSimplifierHandlers.hpp
- [x] compiler/optimizer/OMRSimplifierTableEnum.hpp

**Intermediate Language (IL)**

- [x] compiler/il/ILOpCodeProperties.hpp **no need**
- [x] compiler/il/OMRILOpCodeProperties.hpp
- [x] compiler/il/ILOpCodes.hpp **no need**
- [x] compiler/il/OMRILOpCodesEnum.hpp

**RAS**

- compiler/ras/Tree.cpp -----> in a separate branch (TBD).

**X86/AMD64**

- [x] compiler/x/i386/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/x/i386/codegen/OMRTreeEvaluatorTable.hpp
- [x] compiler/x/amd64/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/x/amd64/codegen/OMRTreeEvaluatorTable.hpp

**ARM**

- [x] compiler/arm/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/arm/codegen/OMRTreeEvaluatorTable.hpp
- [x] compiler/aarch64/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/aarch64/codegen/OMRTreeEvaluatorTable.hpp


**Power**

- [x] compiler/p/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/p/codegen/OMRTreeEvaluatorTable.hpp


**Z**

- [x] compiler/z/codegen/TreeEvaluatorTable.hpp  **no need**
- [x] compiler/z/codegen/OMRTreeEvaluatorTable.hpp

fjeremic commented on compiler/il/OMROpcodes.hpp:
```diff
@@ -0,0 +1,795 @@
+   MACRO(TR::lconst,"lconst",ILProp1::LoadConst,ILProp2::ValueNumberShare,ILProp3::LikeUse,0,TR::Int64,ILTypeProp::Size_8 | ILTypeProp::Integer,ILChildProp::NoChildren,TR::BadILOp,TR::BadILOp,TR::BadILOp,TR::BadILOp) \
+   MACRO(TR::fconst,"fconst",ILProp1::LoadConst,ILProp2::ValueNumberShare,ILProp3::LikeUse,0,TR::Float,ILTypeProp::Size_4 | ILTypeProp::Floating_Point,ILChildProp::NoChildren,TR::BadILOp,TR::BadILOp,TR::BadILOp,TR::BadILOp) \
+   MACRO(TR::dconst,"dconst",ILProp1::LoadConst,ILProp2::ValueNumberShare,ILProp3::LikeUse,0,TR::Double,ILTypeProp::Size_8 | ILTypeProp::Floating_Point,ILChildProp::NoChildren,TR::BadILOp,TR::BadILOp,TR::BadILOp,TR::BadILOp) \
+   MACRO(TR::bconst,"bconst",ILProp1::LoadConst,ILProp2::ValueNumberShare,ILProp3::LikeUse,0,TR::Int8,ILTypeProp::Size_1 | ILTypeProp::Integer,ILChildProp::NoChildren,TR::BadILOp,TR::BadILOp,TR::BadILOp,TR::BadILOp) \
```
Before you get too far into this I'd point out that at least to me the readability of these definitions has regressed. However these would just be formatting changes we could decide on later.

oneturkmen replied:
Thanks for the feedback! Definitely, there is a need for more readability there. I have the following ideas:

1. Since there are more properties to add, I want to turn single-lined macro arguments to multiple lines. Otherwise, the line length may easily exceed 100 chars and it is going to look stuffed and quite messy.

2. If we follow (1), we can actually append a comment to each line so developers reading that `OMROpcodes.hpp` know what each argument means. This could be automatically generated (as I generated `OMROpcodes.hpp` using scripts).

fjeremic replied:
Sounds reasonable. If we can maintain a similar look to what we have in OMRILOpCodeProperties.hpp I think most people will be happy.

fjeremic commented on compiler/il/OMROpcodes.hpp:
```diff
@@ -0,0 +1,11786 @@
+      TR::BadILOp, /* .booleanCompareOpCode */ \
+      TR::BadILOp, /* .ifCompareOpCode */ \
+      sconst, /* load short integer constant (16-bit signed 2's complement) */ \
+   ) \
```
Looking better. Perhaps having the comments on the left side (since they are fixed width and consistent from one opcode to the next) may make things a lot easier to read and parse :

```
   MACRO(\
      /* .opcode               = */ TR::sconst, \
      /* .name                 = */ "sconst", \
      /* .properties1          = */ ILProp1::LoadConst, \
      /* .properties2          = */ ILProp2::ValueNumberShare,  \
      /* .properties3          = */ ILProp3::LikeUse, \
      /* .properties4          = */ 0,  \
      /* .dataType             = */ TR::Int16,  \
      /* .typeProperties       = */ ILTypeProp::Size_2 | ILTypeProp::Integer, \
      /* .childProperties      = */ ILChildProp::NoChildren, \
      /* .swapChildrenOpCode   = */ TR::BadILOp, \
      /* .reverseBranchOpCode  = */ TR::BadILOp, \
      /* .booleanCompareOpCode = */ TR::BadILOp, \
      /* .ifCompareOpCode      = */ TR::BadILOp, \
      /* .description          = */ sconst, \
   ) \
```

oneturkmen replied:
That's what I thought of doing initially. However, some opcodes' last property (enum values from *OMRILOpcodesEnum.hpp*) has a long single-line comment.

Should I disregard them, or have them something like that:

```
      /* .description          = */ sconst, \ // load short integer constant (16-bit signed 2's complement)
```

fjeremic replied:
I don't see that last field in [1] at all. I suggest if there is a comment like the in your previous post you can just include it in the `.description` comment, for example:

```
...
      /* .reverseBranchOpCode  = */ TR::BadILOp, \
      /* .booleanCompareOpCode = */ TR::BadILOp, \
      /* .ifCompareOpCode      = */ TR::BadILOp, \
      /* .description          =    load short integer constant (16-bit signed 2's complement), */ \
```

[1] https://github.com/eclipse/omr/blob/1862ec43d12a8f5d216554e78a591a40726f32e0/compiler/il/OMRILOps.hpp#L42-L63

oneturkmen replied:
Oops, I did not mean properties, but rather the last argument of these macros. You are right, properties end with `.ifCompareOpCode`, and then we have the opcode enum value.

Would it still be a good solution given that these comments can go up to 100 characters? E.g., [this opcode from OMRILOpCodesEnum.hpp](https://github.com/eclipse/omr/blob/master/compiler/il/OMRILOpCodesEnum.hpp#L700)

fjeremic replied:
Yeah, leave them in there for now so we preserve the information. These comments likely belong elsewhere, probably in a per IL documentation with some structure, but we're not there yet. Let's preserve them for now and once we find a good home for them we will eventually remove the comments.

oneturkmen replied:
I think it is a bit better now. What do you think?

```c++
MACRO(\
   /* .opcode               = */ TR::variableNewArray, \
   /* .name                 = */ "variableNewArray", \
   /* .properties1          = */ ILProp1::HasSymbolRef, \
   /* .properties2          = */ ILProp2::CanRaiseException| ILProp2::MayUseSystemStack, \
   /* .properties3          = */ ILProp3::LikeDef, \
   /* .properties4          = */ 0, \
   /* .dataType             = */ TR::Address, \
   /* .typeProperties       = */ ILTypeProp::Reference, \
   /* .childProperties      = */ ONE_CHILD(ILChildProp::UnspecifiedChildType), \
   /* .swapChildrenOpCode   = */ TR::BadILOp, \
   /* .reverseBranchOpCode  = */ TR::BadILOp, \
   /* .booleanCompareOpCode = */ TR::BadILOp, \
   /* .ifCompareOpCode      = */ TR::BadILOp, \
   /* .description          = */ variableNewArray, /* new array - type not known at compile time, type must be a j9class, do not use type enums */ \
)
```

fjeremic replied:
Looks much better!

oneturkmen commented:
So I started working on `optimizer/OMRSimplifierTableEnum.hpp` and stumbled upon [these lines of code](https://github.com/eclipse/omr/blob/2aa19c4f04329039dd4baba88684588274092efb/compiler/optimizer/OMRSimplifierTableEnum.hpp#L197):

```
#ifdef TR_HOST_X86
   dftSimplifier,           // TR::f2i
   dftSimplifier,           // TR::f2l
#else
   f2iSimplifier,
   f2lSimplifier,
#endif
```

I am a bit stuck here because one [cannot use ifdefs inside a macro definition](https://stackoverflow.com/questions/7246512/ifdef-inside-a-macro), so having `#ifdef` inside FOR_EACH_OPCODE definition is not possible.

Alternative (horrible) is to have two definitions of FOR_EACH_OPCODE, i.e.,

```
#ifdef TR_HOST_X86
   FOR_EACH_OPCODE(...) // + all 735 14-line macros
#else
   FOR_EACH_OPCODE(...) // + all 735 14-line macros but where one macro has a different definition
#endif
```

this one is not scalable because I suspect there are such ifdefs in other files (such as `#ifdef J9_SPECIFIC`).

fjeremic commented:
So for these cases we need a 1-1 mapping of opcode to evaluator/simplifier handler or whatever. So If there is an `#ifdef` like that, then reduce it down to a single function, i.e. those lines seem to correpsond to the `TR::f2i` opcode, so you would have a `f2iSimplifier`, and then inside you would do:

```
void Whatever:f2iSimplifier(...)
{
#ifdef TR_HOST_X86
   return dftSimplifier(...);
#else
   ...
#endif
}
```

That is, just migrate any conditionals into the function itself.

oneturkmen commented:
Just to confirm - this is the [function](https://github.com/eclipse/omr/blob/master/compiler/optimizer/OMRSimplifierHandlers.cpp#L12163) where I should integrate `#ifdef` and call other function, namely `dftSimplifier`?

fjeremic commented:
Right. Each IL opcode should have a 1-1 mapping to a function call. For Simplifier these would be simplifier handler functions, for instruction selection these would be evaluator functions, etc.

If there is any conditionals based on platform, architecture, whatever just put the conditional in the function itself.

oneturkmen commented:
The following [file](https://github.com/eclipse/omr/blob/2aa19c4f04329039dd4baba88684588274092efb/compiler/optimizer/OMRSimplifierHandlers.hpp) does not get affected, right? I am not sure why I added it, but asking just in case.

fjeremic commented:
> The following [file](https://github.com/eclipse/omr/blob/2aa19c4f04329039dd4baba88684588274092efb/compiler/optimizer/OMRSimplifierHandlers.hpp) does not get affected, right? I am not sure why I added it, but asking just in case.

I think we should canonicalize the names, i.e. the 1-1 mapping for all simplifier handlers should be named following a pattern, for example `<name-of-IL>Simplifier` if they're not already in that form. Then the macro you're introducting would be trivial to implement for the simplifier handler table. I think we spoke about this on the OMR Architecture call.

It may introduce a ton of stub functions but we agreed that was an ok tradeoff.

andrewcraik commented on compiler/il/OMROpcodes.hpp:
```diff
@@ -0,0 +1,13256 @@
+
+#define FOR_EACH_OPCODE(MACRO) \
+   MACRO(\
+      /* .opcode               = */ TR::BadILOp, \
+      /* .name                 = */ "BadILOp", \
+      /* .properties1          = */ 0, \
+      /* .properties2          = */ 0, \
+      /* .properties3          = */ 0, \
+      /* .properties4          = */ 0, \
+      /* .dataType             = */ TR::NoType, \
+      /* .typeProperties       = */ 0, \
+      /* .childProperties      = */ ILChildProp::NoChildren, \
+      /* .swapChildrenOpCode   = */ TR::BadILOp, \
+      /* .reverseBranchOpCode  = */ TR::BadILOp, \
+      /* .booleanCompareOpCode = */ TR::BadILOp, \
+      /* .ifCompareOpCode      = */ TR::BadILOp, \
+      /* .description          = */ BadILOp = 0, /* illegal op hopefully help with uninitialized nodes */ \
+      /* .simplifierHandler    = */ dftSimplifier, \
+      /* .vpHandler            = */ constrainChildren, \
```
Is there a way the comment field could be picked up by doxygen so it could generate the opcode documentation from the opcode definition?

oneturkmen replied:
@Leonardo2718 both `.opcode`, `.name` and `.description` could be easily substituted by using `.description` (which should rather be named `.enumValue`).

However, there is the first opcode (BadILOp) that I am struggling with because there is a dilemma:

In *OMRILOpCodesEnum.hpp*, we should use `BadILOp = 0`; whereas we need just `BadILOp` for token pasting with simplifier handlers, VP handlers, as well as architecture-specific tree evaluators. If I allow for the former, token pasting results in incorrect macro name, e.g., `BadILOp = 0VPHandler`. If I allow for the latter, a static assert checking `NumILOps` fails.

Leonardo2718 replied:
I'd say you leave it as is this for now. I haven't been able to think of an elegant way of doing this yet.

oneturkmen commented:
@fjeremic as far as I understand, there will be a unique function (simplifier handler) for each opcode.
E.g., 

```
TR::Node * aconstSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s) {
   return dftSimplifier(node, block, s);
}
```

Would not that create a lot of redundant code (for example, `dftSimplifier` is used by 270 opcodes)

oneturkmen commented:
Also, in _compiler/ras/Tree.cpp_, can I just use `.dataType` property specified in the Opcodes.hpp for [the following table](https://github.com/eclipse/omr/blob/master/compiler/ras/Tree.cpp#L2297)?

oneturkmen commented:
Another quick question w.r.t. architecture-specific tree evaluators. Many of them are of the form `TR::TreeEvaluator::<opcode>Evaluator,`.

I was initially thinking of doing token pasting (basically taking an already existing property and pasting it there with suffixes), but this is not going to work because some of the opcodes in some architectures reuse other opcodes' evaluators. In addition, just leaving name of an opcode inside a macro would be confusing.

What do you think about the idea of storing just `<opcode>TreeEvaluator` in order to avoid any confusion (for better readability)?

Thank you.

fjeremic commented:
> Would not that create a lot of redundant code though? (for example, `dftSimplifier` is used by 270 opcodes)

See below.

> I was initially thinking of doing token pasting (basically taking an already existing property and pasting it there with suffixes), but this is not going to work because some of the opcodes in some architectures reuse other opcodes' evaluators. In addition, just leaving name of an opcode inside a macro would be confusing.
> 
> What do you think about the idea of storing just `<opcode>TreeEvaluator` in order to avoid any confusion (for better readability)?

We discussed this on the OMR Architecture call [1] and the general consensus was that we did not want to leak function mappings into the IL macros. So for example we do **not** want to expose `dftSimplifier` in the definition of the IL because that would introduce a coupling between the IL definition and some use of the IL in some table.

Instead the consensus was that the IL definition macros _only_ describe the IL, and that information is used locally in whatever table to generate a function mapping, so for example VP, evaluators, simplifier handlers, etc. So yes, that means that the names _must_ be canonical, and perhaps overly redundant in case a user wishes to create a visitor pattern pass which only affects a few IL, however on the call we felt that trade-off was worth while.

To explicitly answer your question, your above thinking is correct. We would use the IL definion macro to extract the name of the IL and then we would add a suffix to generate the function name which that IL maps to. For example, if we are generating evaluators then just as you described we would generate a 1-1 mapping between IL and an evaluator function `<opcode>Evaluator`. If we are generating simplifier handlers then we map each IL to a function called `<opcode>SimplifierHandler` or something like that.

Now of course, unlike evaluators, not ever IL participates in simplification, so there will be a lot of redundancy, where for example we may have:

```
TR::Node* iandSimplifierHandler(...)
{
    return andSimplifierHelper(...);
}

TR::Node* landSimplifierHandler(...)
{
    return andSimplifierHelper(...);
}

// etc...
``` 

The consensus on the architecture call was that this duplication and boilerplate is an ok trade-off, because otherwise we would have to leak simplifier handler function names into the IL definitions.

[1] https://www.youtube.com/watch?v=21yPv8GsvY4&feature=youtu.be

Leonardo2718 commented:
Since the OMR Architecture call where this was discussed, I've come up with a few alternative solutions for dealing with handlers and evaluators. The goal was to reduce the amount of duplication without having to specify the function names right in the macros. For completeness, I will summaries three solutions that I thought might be worth considering, all with varying levels of ugliness.

### 1. A function templated on the opcode enum and specialized for the different opcodes

This is the most straightforward solution, IMHO.

Here, there would be one template function that takes an opcode enum value as template parameter. Template specializations of the function would then be provided for different opcodes as needed. If no template specialization is defined, the unspecialized default implementation is used.

```c++
/*** OMRSimplifierHandlers.hpp ***/

// default simplifier handler
template <TR::ILOpCodes OP>
TR::Node* simplifierHandler(...)
   {
   return dftSimplifier(...);
   }

// sepcialized simplifier
template <>
TR::Node* simplifierHandler<TR::iand>(...)
   {
   return andSimplifierHelper(...);
   }

/*** OMRSimplifierTableEnum.hpp ***/

#define GET_SIMPLIFIER_HANDLER(\
   opcode, \
   name, \
   prop1, \
   prop2, \
   prop3, \
   prop4, \
   dataType, \
   typeProps, \
   childProps, \
   swapChildrenOpcode, \
   reverseBranchOpcode, \
   boolCompareOpcode, \
   ifCompareOpcode, \
   enumValue, \
   ...) simplifierHandler<opcode>,

FOR_EACH_OPCODE(GET_SIMPLIFIER_HANDLER)

#undef GET_SIMPLIFIER_HANDLER
```

One drawback of this solution is that the compiler would duplicate the definition of the unspecialized function for every opcode it's used for, even if the function body doesn't depend on the opcode and it's identical for all of them.

### 2. A class template with a static member function

This solution is similar to the previous one in that it also relies on template specialization. However, it solves the problem of duplicate function bodies by using class templates instead of function templates.

Here, the default implementation of the function is a static member of a *non-template* base class. A template class extends the base class but overrides/shadows nothing. Then, the template class can be specialized to provide a different static method for different opcodes.

```c++
/*** OMRSimplifierHandlers.hpp ***/

// non-template base class with the default handler
class SimplifierHandlerBase
   {
   public:
   static TR::Node* handler(...)
      {
      return dftSimplifier(...);
      }
   };

// template class used to access the default handler
template <TR::ILOpCodes OP>
class SimplifierHandler : public SimplifierHandlerBase {};

// specialization with a custom handler
template <>
class SimplifierHandler<TR::iand>
   {
   public:
   static TR::Node* handler(...)
      {
      return andSimplifierHelper(...);
      }
   };

/*** OMRSimplifierTableEnum.hpp ***/

#define GET_SIMPLIFIER_HANDLER(\
   opcode, \
   name, \
   prop1, \
   prop2, \
   prop3, \
   prop4, \
   dataType, \
   typeProps, \
   childProps, \
   swapChildrenOpcode, \
   reverseBranchOpcode, \
   boolCompareOpcode, \
   ifCompareOpcode, \
   enumValue, \
   ...) SimplifierHandler<opcode>::handler,

FOR_EACH_OPCODE(GET_SIMPLIFIER_HANDLER)

#undef GET_SIMPLIFIER_HANDLER
```

All the non-specialized classes will use the same function because there is only one instance of the base class even if multiple instances of the derived template class are produced by the compiler.

### 3. `#define` handlers to the default handler

I do not like this solution at all but I've included it here because it does not rely on templates, and I know not everyone is comfortable with templates.

This solution uses the same approach previously discussed in this PR of using token pasting to produce the name of each handler. However, instead of implementing functions that delegate to a  default handler, the name of each function is `#define`d to the name of the default handler.


```c++
/*** OMRSimplifierHandlers.hpp ***/

TR::Node* dftSimplifier(...)
   {
   //...
   }

#define aconstSimplifierHandler dftSimplifier

/*** OMRSimplifierTableEnum.hpp ***/

#define ID(x) x // needed to force `x` to be macro expanded

#define GET_SIMPLIFIER_HANDLER(\
   opcode, \
   name, \
   prop1, \
   prop2, \
   prop3, \
   prop4, \
   dataType, \
   typeProps, \
   childProps, \
   swapChildrenOpcode, \
   reverseBranchOpcode, \
   boolCompareOpcode, \
   ifCompareOpcode, \
   enumValue, \
   ...) ID(enumValue ## simplifierHandler), 

FOR_EACH_OPCODE(GET_SIMPLIFIER_HANDLER)

#undef GET_SIMPLIFIER_HANDLER
```

### 4. ...

There are, of course, other possible solutions but these three were the closest to the current, overall approach.

oneturkmen commented:
Thanks @Leonardo2718 for a good comment. Let's try to see how solution no.3 works (with `#define`s).

Btw should ValuePropagation handler get processed in the same manner, i.e., with one-to-one mapping?

oneturkmen commented:
@fjeremic @Leonardo2718 I finished all the files except *compiler/ras/Tree.cpp*, which I am going to change in a separate branch. Could you please review the changes? Thank you.

fjeremic commented:
@oneturkmen could you rebase to fix the conflicts? Seems the automated builds failed the last time they ran. We should get clean builds before we start looking for reviews. Thanks for finishing this though!

0xdaryl commented:
Just re-iterating the previous request for rebasing to resolve the merge conflicts.

oneturkmen commented:
My apologies for the late reply/resolution. The conflict is resolved, but there is still a problem with the first opcode in one of the files https://github.com/eclipse/omr/pull/4915#discussion_r407118477.

oneturkmen commented:
@fjeremic @0xdaryl @Leonardo2718 I fixed some of the merge conflicts, as well as the problem caused by BadILOp's ordering. However, now I experience the following error:

```bash
46: *** VALIDATION ERROR ***
46: Node: iadd n4n
46: Method: file:line:name
46: Child 1 has unexpected type Int16 (expected Int32)
46: 
46: *** VALIDATION ERROR ***
46: Node: sadd n4n
46: Method: file:line:name
46: Child 0 has unexpected type Int32 (expected Int16)
46: 
46: *** VALIDATION ERROR ***
46: Node: aiadd n4n
46: Method: file:line:name
46: Child 1 has unexpected type Int64 (expected Int32)
46: 
```

One of the tests, namely "ILValidatorTest/IllformedTrees", fails. I suspect the issue may be due to the ordering in `il/OMRILOpCodeProperties.hpp` file. I double-checked it, as well as the commits that changed this file since last time. Also double-checked the tree evaluator tables in all architectures, everything looks good. Do you have any ideas why this might occur?

P.S. Also, refactored RISC-V tree evaluator table, which seems to look good (does not cause any build errors).

Leonardo2718 commented:
@oneturkmen I don't think the `ILValidatorTest/IllformedTrees` tests are the problem (those messages are normal). But if you look at https://dev.azure.com/eclipse-omr/omr/_build/results?buildId=2131&view=logs&jobId=7f33e5bd-7764-5d8a-ba2e-506e078b9c3f&j=7f33e5bd-7764-5d8a-ba2e-506e078b9c3f&t=dba14c36-22cc-56db-9df9-69f31457753b
it looks like there is a problem with an evaluator:

```
46: [----------] 20 tests from LogicalTest/Int16LogicalUnary
46: Assertion failed at /home/vsts/work/1/s/compiler/codegen/OMRTreeEvaluator.cpp:585: false
46: 	Opcode sbyteswap is not implemented
```

fjeremic commented:
@genie-omr build all

fjeremic commented:
@genie-omr build all

fjeremic commented:
@genie-omr build all

charliegracie commented:
@genie-omr build all

AdamBrousseau commented:
@genie-omr build all

AdamBrousseau commented:
@genie-omr build all

AdamBrousseau commented:
genie-omr build all

AdamBrousseau commented:
genie-omr build all

AdamBrousseau commented:
genie-omr build all

0xdaryl commented:
@oneturkmen : genie builds launched and there appear to be some real code failures you need to investigate.

oneturkmen commented:
@0xdaryl it looks like I accidentally set `dselectEvaluator` for `dselect`, but should have been `fselect`. Fixed now. Thanks!

fjeremic commented:
@genie-omr build all

oneturkmen commented:
Looks like there's a problem with ccache: https://dev.azure.com/eclipse-omr/omr/_build/results?buildId=2204&view=logs&j=196c4937-6990-58b5-7d18-234ac2f0888b&t=255e13c9-34e2-4532-9445-c05f95b80daa

Is it something on my end that caused it?

fjeremic commented:
> Is it something on my end that caused it?

Seems like an infrastructure issue in Azure. I've restarted it. We're going to need a few of us to comb through this as it looks ready.

fjeremic commented on compiler/z/codegen/OMRTreeEvaluatorTable.hpp:
```diff
@@ -24,780 +24,774 @@
-   TR::TreeEvaluator::bxorEvaluator,        // TR::bxor
-   TR::TreeEvaluator::sxorEvaluator,        // TR::sxor
 
 #define SIGNED true
 #define UNSIGN false
```
It seems odd now that these are present. How does the macro expansion work inside of other macros? I'm a bit worried with the layers of macros this now adds complexity. I think we should just inline these values in the 6 or so places that use it and get rid of these `#define` and `#undef` statements.

oneturkmen replied:
The macro first expands the SIGN and UNSIGN uses, turning it to `true` and `false`, respectively. It's about 13 SIGN and 6 UNSIGN uses. But I agree, inlining the two macros would reduce the complexity.

fjeremic replied:
I propose we fold them in this PR to simplify things.

fjeremic commented on compiler/z/codegen/OMRTreeEvaluatorTable.hpp:
```diff
@@ -24,780 +24,774 @@
+   boolCompareOpcode, \
+   ifCompareOpcode, \
+   enumValue, \
+   ...) enumValue ## ZEvaluator,
```
Is there a functional reason why we had to include the arch in the name? It is less useful now that we have for example `iloadZEvaluator` and `iloadPowerEvaluator` as we can't search of all the evaluators on all platforms easily and compare them, which is something often done.

IMO the goal should be to be as similar across platforms as possible. The platform specific code is already in its own directory for example `/z/...` and `/p/...`.

oneturkmen replied:
That's a good point. Should we discuss this more, or can I proceed with refactoring?

fjeremic replied:
Yes we should discuss. Let me tag a few people here @0xdaryl @aviansie-ben @Leonardo2718. Thoughts?

Leonardo2718 replied:
I agree that including the arch in the evaluator name doesn't provide that much value. I think dropping it makes sense.

fjeremic replied:
@0xdaryl @aviansie-ben any other thoughts?

0xdaryl replied:
I agree the architecture name shouldn't be there.

fjeremic replied:
I think we have enough of a consensus to revert back the names without the platform suffixes and make them uniform across codegens.

oneturkmen replied:
It looks like there is a naming collision if I remove architecture name from the macro name, i.e., `#define iloadEvaluator TR::...::iloadEvaluator` probably re-defines (and re-expands) when a macro is called.

Do you think having underscores for the macro names would be OK? I.e., smth like `#define _iloadEvaluator TR...`

fjeremic replied:
That seems reasonable to me.

fjeremic commented on compiler/arm/codegen/FPTreeEvaluator.cpp:
```diff
@@ -3138,4 +3138,14 @@ TR::Register *OMR::ARM::TreeEvaluator::fRegStoreEvaluator(TR::Node *node, TR::Co
    return NULL;
    }
 
+TR::Register *OMR::ARM::TreeEvaluator::fmaxEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return NULL; 
+   }
+
+TR::Register *OMR::ARM::TreeEvaluator::fminEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return NULL; 
+   }
```
These seem to be mapped already in the above define sequence:

```
#define fmaxARM64Evaluator TR::TreeEvaluator::unImpOpEvaluator 
#define dmaxARM64Evaluator TR::TreeEvaluator::unImpOpEvaluator 
```

Do we need these two functions then?

oneturkmen replied:
I think [*compiler/arm/codegen/OMRTreeEvaluatorTable.hpp*](https://github.com/eclipse/omr/blob/f9ef03b68a474ae911b9dfac8e6e59c8c5c088b8/compiler/arm/codegen/OMRTreeEvaluatorTable.hpp#L710) contains:

```cpp
#define fmaxARMEvaluator TR::TreeEvaluator::fmaxEvaluator
#define dmaxARMEvaluator TR::TreeEvaluator::fmaxEvaluator
```

And we also have two (mutually exclusive) method definitions based on `#if (defined(__VFP_FP__) && !defined(__SOFTFP__))` on [Line 40](https://github.com/eclipse/omr/blob/f9ef03b68a474ae911b9dfac8e6e59c8c5c088b8/compiler/arm/codegen/FPTreeEvaluator.cpp#L40)

fjeremic replied:
Right! This is the ARM evaluators, and the defines were in ARM64. I made a mistake thinking they were in the same file. I rescind my review comment. All is well!

fjeremic commented:
@oneturkmen thanks for addressing that quickly. It seems a conflict was just introduced in RISC-V codegen. While you address that I'm going to take one hard look at the whole PR again. If there are no major issues I propose we merge it now to keep the ball rolling.

I'm going to test this PR in OpenJ9 as well to ensure we don't break anything there. Once again many thanks for completing this work. We are almost at the finish line!

fjeremic commented on compiler/il/OMROpcodes.hpp:
```diff
@@ -0,0 +1,11395 @@
+#ifndef OMR_OPCODEMACROS_INCL
+#define OMR_OPCODEMACROS_INCL
+
+/*#ifdef FIRST_OPCODE_INIT
+#define FIRST_OPCODE_VAL BadILOp=0
+#else
+#define FIRST_OPCODE_VAL BadILOp
+#endif*/
```
Do we need this? If not let's remove it.

oneturkmen replied:
Done

fjeremic commented on compiler/il/OMROpcodes.hpp:
```diff
@@ -0,0 +1,11395 @@
+   MACRO(\
+      /* .opcode               = */ TR::iflcmpgt, \
+      /* .name                 = */ "iflcmpgt", \
+      /* .properties1          = */ ILProp1::BooleanCompare | ILProp1::Branch | ILProp1::TreeTo
```
Similarly here.

oneturkmen replied:
Same here (I believe this relates to the #endif, commented out at the end of the file)

fjeremic commented:
@oneturkmen just did another final round. Only found those two minor things to clean up. We'll also need to squash the 0c17f76 merge commit. Otherwise I think we're good to go. I'm going to pull in your changes now and get a build going with OpenJ9.

fjeremic commented:
Finished testing. Looks good. Last thing is to clean up the merge commit and those two reviews up above and we can merge.

fjeremic commented:
@genie-omr build all

fjeremic commented:
Once again, many thanks for this work @oneturkmen. Congrats on merging such a large PR!

oneturkmen commented:
Thanks a lot for the feedback and code reviews! :tada:
